cpp/include/cudf/fixed_point/fixed_point.hpp

```diff
@@ -121,6 +121,30 @@ CUDA_HOST_DEVICE_CALLABLE constexpr T right_shift(T const& val, scale_type const
   return val / ipow<Rep, Rad>(scale._t);
 }
 
+/** @brief Function that performs a rounding `right shift` scale "times" on the `val`
+ *
+ * The scaled integer equivalent of 0.5 is added to the value before truncating such that
+ * any remaining fractional part will be rounded away from zero.
+ *
+ * Note: perform this operation when constructing with positive scale
+ *
+ * @tparam Rep Representation type needed for integer exponentiation
+ * @tparam Rad The radix which will act as the base in the exponentiation
+ * @tparam T Type for value `val` being shifted and the return type
+ * @param val The value being shifted
+ * @param scale The amount to shift the value by
+ * @return Shifted value of type T
+ */
+template <typename Rep, Radix Rad, typename T>
+CUDA_HOST_DEVICE_CALLABLE constexpr T right_shift_rounded(T const& val, scale_type const& scale)
+{
+  constexpr Rep base = static_cast<Rep>(Rad);
+  Rep const factor   = ipow<Rep, Rad>(scale._t - 1);
+  Rep const half     = (base / 2) * factor;
```

Comment:
This would be incorrect if the Radix were allowed to be an odd number. Probably unlikely, but I suppose I could add a static_assert or something to be safe.

Reply:
Probably an `assert` and an `ASSERT_DEATH` unit test would be good.

Reply:
For odd radixes, 1/2 is irrational so rounding would require adding a repeating number (`base / 2` rounded down, repeating). e.g. in base 5 decimal, 1.2222 should round down while 1.2223 should round up. 5/2 repeating is .2222, so 1.2222 + .2222 = 1.4444 (truncates to 1) and 1.2223 + .2222 = 2.

The base is a compile time parameter and Radix is currently only defined for base 10 and base 2, so I'd rather add a `static_assert` than a runtime `assert`. I don't think I can add a unit test for `static_assert`, but I'll test locally to make sure it fails to compile for odd Radix.

However, if needed we can add a specialization for odd radixes that builds up the repeating fraction in the same loop as ipow. I'll reference this comment next to the assert so the solution is available.
```cpp
template <typename Rep, Radix Rad, typename T,
    std::enable_if_t<(Rad % 2 == 1, void*> = nullptr> // enable for odd Radix
CUDA_HOST_DEVICE_CALLABLE constexpr T right_shift_rounded(T const& val, scale_type const& scale)
{
  constexpr Rep base = static_cast<Rep>(Rad);
  Rep factor = 1;
  Rep half = 0;
  for (Rep i = 0; i < scale._t; ++i) {
    half += (base / 2) * factor;
    factor *= base;
  }
  return (val >= 0 ? val + half : val - half) / factor;
}
```

Reply:
Scratch that, you can just use `((base + 1) / 2) - 1` to compute the repeating fraction. :facepalm:

Reply:
Ok, it seemed like there should be a simpler way... and duh, `half` can just be `ipow<Rep, Rad>(scale._t) / 2`, which works for both even and odd radixes and simplifies the code a bit.